### DISCUSSION TITLE
How can I get custom directive on input field?

### DISCUSSION THREAD

**Discussion by maxixcom** [Netflix/dgs-framework#617] [Q&A] [answered]
Hello, is it possiable with dgs to check if there is a custom directive is set to input field?

Schema example:

```graphql
directive @hasRole(role: String!) on INPUT_FIELD_DEFINITION

mutation {
    updateUser(id ID!, input: UpdateUserInput)  :  User
}

type User {
    id: ID!
    name: String
    role: String
}

input UpdateUserInput {
    name: String
    role: String @hasRole("admin")
}



```

Thank you in advance.

**Comment by srinivasankavitha** [comment]
There is no special logic in the DGS framework that is needed to support this functionality. If it is possible to do this per spec (and seems to be the case here: https://spec.graphql.org/June2018/#sec-Language.Directives), it should work. That said, I have not tried this myself. I assume you will need be adding specific logic to handle the directive and so you will need to check if you have everything you need at that point.

**Reply by maxixcom** [reply]
Thank you for your reply. Actually I'm new to java `graphql-java` and `dgs`, and unfortunately I couldn't find any example on this case. I have no idea where I should put my logic to check directives on input fields.
Could you give me some clue, where I should start with adding logic to test the directives?

**Comment by srinivasankavitha** [comment]
Here is a good reference describing how to set up custom directives:
https://www.graphql-java.com/documentation/v11/sdl-directives/

**Reply by maxixcom** [reply]
Thanks for the link. But it is all about directives that are set on type fields (FIELD_DEFINITION).
It's pretty clear how to do it - just wrap one datafetcher with a new one where we can put additional behavior.

The problem is I couldn't apply this approach to input fields (INPUT_FIELD_DEFINITION).
I understand that I have to override `onInputObjectField` method on my Custom Directive class like:



```kotlin

class HasRoleDirective : SchemaDirectiveWiring {
    override fun onInputObjectField(environment: SchemaDirectiveWiringEnvironment<GraphQLInputObjectField>): GraphQLInputObjectField {
        // Put some code that will be run when input object is passed

        // How to do this?

    }
}

```

Maybe I'm wrong, but I think that there are no datafetchers for input objects and nothing to be wrapped as with regular fields. So how can I dynamically transform my input object fields?

**Comment by srinivasankavitha** [comment]
I assume you will have to do the transformation within the data fetcher for
the query that uses the input type? Again, I haven't tried this since
myself, but I assume you have access to the input arguments and can use
that to do the transformation. If not, it's probably not supported by
graphql-java to begin with.

**Reply by maxixcom** [reply]
I wish to do it in general way, with SchemaDirectiveWiring, but checking it in datafetcher would work too.  I just couldn't find a way how to check it in datafetcher whether input field has directive on it or it doesn't  :(

Thank you very much for trying help.

**Comment by srinivasankavitha** [comment]
I am not sure I understand the actual use case. That said, I just I tried overriding the `onInputObjectField` method and that seems to get invoked each time you have a directive set up on the input field. E.g.
```
public class TestDirective implements SchemaDirectiveWiring {
    @Override
    public  GraphQLInputObjectField onInputObjectField(SchemaDirectiveWiringEnvironment<GraphQLInputObjectField> environment) {
        GraphQLInputObjectField field = environment.getElement();
        GraphQLFieldsContainer parentType = environment.getFieldsContainer();
        
       // add necessary transformation logic for the input field here
        return field.transform(builder -> builder
                .defaultValue(field)
        );
    }

}
```
Is that what you are trying to do? What is the objective of setting the directive on the input field?

**Reply by maxixcom** [reply]
Thank you for your reply. It seems the builder in your example run only on start for dynamicaly change schema. I added an example based on `dgs-example-kotlin`:

https://github.com/maxixcom/dgs-examples-kotlin/tree/test-inputobject-directive

I wanted added directive that should reset input field value if the value equals to test value in schema.
But for start, the code just set default value with current time:

https://github.com/maxixcom/dgs-examples-kotlin/blob/a7e5d137c17003c529b67ca068ff473e661d12f5/src/main/kotlin/com/example/demo/directives/ResetValueDirective.kt#L35


So expected behavior - every time you run query like the next one:

```graphql
mutation {
  addReview(
    review: {
      showId: 1
      starScore: 3
    }
  ) {
    username
  }
}
```

you would get review with username equals to `guest-<Time when it's created>`:
```
Run first time
username = Guest-2021-09-15T11:15:45

Run next time
username = Guest-2021-09-15T11:15:52

...
```
But since it run only on start you get always the value it that is built on server start.


> Is that what you are trying to do? What is the objective of setting the directive on the input field?

The original idea is to protect some input fields from changing if user permissions is not enough for it. For example you can modify user like:
```graphql
input UserUpdateInput {
    id: ID!
    firstName: String 
    lastName: String 
    roles: [UserRole!] @hasPermission(permission: "admin:write")
    status: UserStatus @hasPermission(permission: "manager:write")
}

userUpdate(input: UserUpdateInput!) : UserUpdatePayload
```

So you can change any fields but not roles if you haven't `admin:write` permission, the same is for status but having `manager:write` permission is enough to proceed with change.

Of course  I can do all this in datafetcher where I process updating user request, but I wanted to make it universal with directive on input fields.

I hope I can make the idea more clear.

Thank you again.

**Reply by pratgoyal** [reply]
> I am not sure I understand the actual use case. That said, I just I tried overriding the `onInputObjectField` method and that seems to get invoked each time you have a directive set up on the input field. E.g.
> 
> ```
> public class TestDirective implements SchemaDirectiveWiring {
>     @Override
>     public  GraphQLInputObjectField onInputObjectField(SchemaDirectiveWiringEnvironment<GraphQLInputObjectField> environment) {
>         GraphQLInputObjectField field = environment.getElement();
>         GraphQLFieldsContainer parentType = environment.getFieldsContainer();
>         
>        // add necessary transformation logic for the input field here
>         return field.transform(builder -> builder
>                 .defaultValue(field)
>         );
>     }
> 
> }
> ```
> 
> Is that what you are trying to do? What is the objective of setting the directive on the input field?

Yes the `onInputObjectField` is triggered on every input field but it does not let us wrap the data fetcher like the `onField` does. 

A `SchemaDirectiveWiringEnvironment` env is passed in to both `onInputObjectField` &  `onField` and the `env.setFieldDataFetcher` returns a `GraphQLFieldDefinition`.  `onInputObjectField` returns a `GraphQLInputObjectField` and not a `GraphQLFieldDefinition` object like `onField`.

Hence, the `onInputObjectField` does not let you return the modified data fetcher. I guess this is something that should be supported by `graphql-java`?

**Comment by srinivasankavitha** [comment, marked as answer]
Actually the validation library provided by graphql-java does what you are looking for: https://github.com/graphql-java/graphql-java-extended-validation/blob/master/src/main/java/graphql/validation/schemawiring/ValidationSchemaWiring.java#L34

You could probably use that approach. Essentially, you will need to override the `onField` method in SchemaDirectiveWiring and replace the data fetcher. Unfortunately, you will need to dig into the data fetching environment to extract the input value definitions and check which ones have the directive applied to them to customize the logic. Don't believe there is an easier way around this.

The extended-validation library does something similar and builds a validation environment with all the necessary information from the DataFetchingEnvironment.

Hope that provides some pointers.

**Reply by maxixcom** [reply]
Yes, I think it's the only option to reach the goal :(
Thanks alot.

**Reply by khacsinhcs** [reply]
@maxixcom I faced with the same issue, what is your final solution? `onField` isn't trigged for field in input, then I can create custom `dataFetcher` for it
